clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -268,72 +279,78 @@ public void poll(long timeout) throws IOException {
         currentTimeNanos = endSelect;
         this.sensors.selectTime.record(endSelect - startSelect, time.milliseconds());
 
-        if (readyKeys > 0) {
-            Set<SelectionKey> keys = this.nioSelector.selectedKeys();
-            Iterator<SelectionKey> iter = keys.iterator();
-            while (iter.hasNext()) {
-                SelectionKey key = iter.next();
-                iter.remove();
-                KafkaChannel channel = channel(key);
-
-                // register all per-connection metrics at once
-                sensors.maybeRegisterConnectionMetrics(channel.id());
-                lruConnections.put(channel.id(), currentTimeNanos);
-
-                try {
-                    /* complete any connections that have finished their handshake */
-                    if (key.isConnectable()) {
-                        channel.finishConnect();
+        if (readyKeys > 0 || !immediatelyConnectedKeys.isEmpty()) {
+            pollSelectionKeys(this.nioSelector.selectedKeys());
+            pollSelectionKeys(immediatelyConnectedKeys);
+        }
+
+        addToCompletedReceives();
+
+        long endIo = time.nanoseconds();
+        this.sensors.ioTime.record(endIo - endSelect, time.milliseconds());
+        maybeCloseOldestConnection();
+    }
+
+    private void pollSelectionKeys(Iterable<SelectionKey> selectionKeys) {
+        Iterator<SelectionKey> iterator = selectionKeys.iterator();
+        while (iterator.hasNext()) {
+            SelectionKey key = iterator.next();
+            iterator.remove();
+            KafkaChannel channel = channel(key);
+
+            // register all per-connection metrics at once
+            sensors.maybeRegisterConnectionMetrics(channel.id());
+            lruConnections.put(channel.id(), currentTimeNanos);
+
+            try {
+
+                /* complete any connections that have finished their handshake */
+                if (key.isConnectable()) {
```

Comment:
If the key came from immediatelyConnectedKeys then this will be false and the clause not entered. You could also test for selectionKeys == immediatelyConnectedKeys but that'd be brittle. If the project was java 8 I'd add a BooleanSupplier param and do:

`pollSelectionKeys(this.nioSelector.selectedKeys(), key -> key.isConnectable());
pollSelectionKeys(immediatelyConnectedKeys, key -> true);`

... but, since it appears you're still Java 7 then I'd add a forceConnect boolean param.